### ISSUE TITLE
Bug: broadcasting should be avoided in Dense

### ISSUE THREAD

**Issue by johnnychen94** [FluxML/Flux.jl#354]
When I read the [Basics](http://fluxml.ai/Flux.jl/stable/models/basics.html) part of Flux tutorials, I found a behavior out of my expectation, that scalar value is accidentally broadcasted in the calculation, and hence output useless values without warning or error.

```julia
using Flux
# model input should be 10-element array
model = Chain(
  Dense(10, 5, σ),
  Dense(5, 2),
  softmax)

model(randn(10)) # 2-element array <----- expected
model(randn(1)) # raise DimensionMismatch <---- expected
model(1) # Tracked 2×10 Array <---- unexpected behavior, should be avoid as far as I concerned
```

I'm new to Flux, so I'm not sure if this is a feature or not. And I'm not sure if other layers has this kind of issues

**Comment by MikeInnes** [comment]
We can just add an `AbstractArray` type restriction to dense to fix this.

**Comment by johnnychen94** [comment]
@MikeInnes 
Well I simply add an `(a::Dense)(x::Number) = a([x]) # prevent broadcasting of scalar` in my PR.

I'm not sure that I understand what you mean. Is that changing Dense from
```julia
struct Dense{F,S,T}
  W::S
  b::T
  σ::F
end
```
to
```julia
struct Dense{F, S <: AbstractArray,T <: AbstractArray}
  W::S
  b::T
  σ::F
end
```

**Comment by MikeInnes** [comment]
Right, the `Number` method isn't necessary though, as adding the `AbstractArray` type hint will mean that you get a method error.

**Comment by MikeInnes** [comment]
Technically I guess, but that's a very weird use case, and you could use a Diagonal just as easily. I would prefer to give a better error message here and let people do something else for this.
